docs/source/app-dev/bindings-java/codegen.rst

```diff
@@ -380,3 +390,99 @@ The Java code generated for this variant is:
+The Java Code Generator uses Java Generic types to represent the DAML parameterized type.
+
+Below is a DAML fragment defining the parameterized type `Attribute` for use by the `
+
+.. code-block:: daml
+  :caption: Com/Acme.daml
+
+  daml 1.2
+  module Com.Acme where
+
+  data Attribute a = Attribute
+      with v : a
+
+  data BookAttributes = BookAttributes with
+     pages : (Attribute Int)
+     authors : (Attribute [Text])
+     title : (Attribute Text)
+
+A file Java file is generated for the `Attribute` data type that defines the Java Generic class:
+
+Note: If the parameterized type is contained in a type where the *actual* type is specified (as in the `BookAttributes` type above), then the serialization
+and deserialization of the enclosing type provides the necessary methods for serialization and deserialization.
+
+.. code-block:: java
+  :caption: com/acme/Attribute.java
+  :emphasize-lines: 3,8,10
+
+  package com.acme;
+
+  public class Attribute<a> {
+    public final a value;
+
+    public Attribute(a value) { /* ... */  }
+
+    public Record toValue(Function<a, Value> toValuea) { /* ... */ }
+
+    public static <a> Attribute<a> fromValue(Value value$, Function<Value, a> fromValuea) { /* ... */ }
+  }
+
+**Serializing**
+
+To serialize an instance of the `Attribute<a>` data type function for creating the Ledger API equivalent of the attribute value
+is passed to the `toValue` method as the `fromValuea` argument (see the above `com/acme/Attribute.java` source extract).
+
+Below is a Java fragment to serialize an attribute with a `java.lang.Long` value to Ledger API representation using the *method reference*
+`Int64::new` to create a new instance of the Java Bindings value type.
+
+.. code-block:: java
+
+  Attribute<Long> pagesAttribute = new Attributes<>(42L);
+
+  Value serializedPages = pagesAttribute.toValue(Int64::new);
+
+See :ref:`DAML To Java Type Mapping <daml-codegen-java-primitive-types>` for the Java Bindings value types need to be created in the `fromValue` method.
+
+Note: That if the DAML type is a record that has more that one parameterized field, then a function for creating the
+Java Binding values must be supplied for *each* such fields.
+
+**Deserializing**
+
+The deserialization method `fromValue` requires in a similar manner, a function to convert the Java Bindings value types to it's corresponding Java type.
+
+.. code-block:: java
+
+  Attribute<Long> pagesAttribute = Attribute.<Long>fromValue(serializedPages,
+      f -> f.asInt64().getOrElseThrow(() -> throw new IllegalArgumentException("Expected Int field").getValue());
+
+See Java Bindings `Value`_ class for the methods to transform the Java Bindings types into corresponding Java types.
+
+.. _Value: https://docs.daml.com/app-dev/bindings-java/javadocs/com/daml/ledger/javaapi/Value.html
+
+**Parameterized product types**
+
+The serialization of parameterized product types require a multiple stage conversion function where the elements must be
```

Comment:
I don't quite understand this sentence. Are you referring to a record that has a field that is itself a parameterized type? In that case this section is more about *nested parameterized types* I  think.

Reply:
It's more about DamlList and DamlOptional being creating using pre-converted Java Binding values.

Reply:
Hm, if this is about DamlList and DamlOptional, maybe change the heading to refer to those. As it is, DamlList and DamlOptional aren't product types at all.

Reply:
Done